Added Kind to HACKING.md and added HACKING-cli-image.md (#2713)
strimzi/strimzi-kafka-operator: HACKING.md

## Build Pre-Requisites

To build this project you must first install several command line utilities and a Kubernetes or OpenShift cluster.

> You may want to try to use a ready-made Docker image based on CentOS 7 with all the tools pre-installed to quickly perform a build. See the instructions [here](HACKING-cli-image.md).

Reviewer: This give the idea the Kubernetes or OpenShift clusters are included which is not true. MAybe you can split the first sentence into two parts and make it more clear what does this offer.

## Building Strimzi

In order to build Strimzi images you need a running Docker daemon (either local or remote).

> If you also want to run tests or deploy Strimzi CRDs and the Strimzi Cluster Operator you need access to Kubernetes API server.

Reviewer 2: ```suggestion
If you also want to run tests or deploy Strimzi CRDs and the Strimzi Cluster Operator you need access to a Kubernetes API server.
```

The images will then be built and stored in the cluster VM's local image store and then pushed to your configured Docker registry. 

> #### Local build with push to Docker registry used by Kind

Reviewer 2: ```suggestion
#### Local build pushing to the Docker registry used by Kind
```

> When developing locally you might want to push the docker images to the docker repository running as container in your Docker daemon and used by your Kind cluster deployed in the same Docker daemon. This can be quicker and more convenient than pushing to Docker Hub and works even without a network connection.

Reviewer 2: ```suggestion
When developing locally you might want to push the docker images to the docker repository running as a container in your Docker daemon and used by your Kind cluster deployed in the same Docker daemon. This can be quicker and more convenient than pushing to Docker Hub and works even without a network connection.
```

1. Determine the IP where your Docker Registry can be reached from both your local environment and from docker containers.

   When deploying Docker Registry container to Docker it will most likely be insecure (accessed over 'http://' url).
   There are networking differences with different Docker deployments on different local environments.

   On Linux:

       export REGISTRY_IP=$(ifconfig docker0 | grep 'inet ' | awk '{print $2}') && echo $REGISTRY_IP

   On MacOS:

       export REGISTRY_IP=$(ifconfig en0 | grep 'inet ' | awk '{print $2}') && echo $REGISTRY_IP

2. Deploy the Docker Registry if it has not been deployed yet

       export REGISTRY_NAME=docker-registry
       export REGISTRY_PORT=5000

       docker run -d --restart=always -p "$REGISTRY_PORT:$REGISTRY_PORT" --name "$REGISTRY_NAME" registry:2

>    Using `-p` exposes the registry on port 5000 on localhost, but also on the network interface with $REGISTRY_IP address, making it available from your local host as well as from Docker daemon running in a VM, and also from Docker containers running inside Docker daemon (such as Kind cluster running as `kind-control-plane` container).

Reviewer 2: ```suggestion
   Using `-p` exposes the registry on port 5000 on localhost, but also on the network interface with $REGISTRY_IP address, making it available from your local host as well as from the Docker daemon running in a VM, and also from Docker containers running inside the Docker daemon (such as the Kind cluster running as the `kind-control-plane` container).
```

3. Configure your Docker daemon to trust your Docker Registry.

   The way we deployed the Docker Registry here is without configuring `https`, keystores, and certificates. Such a registry is considered `insecure`, and requires additional configuration of Docker daemon, and Kind cluster.
   
   On Linux there is usually a daemon config file at `/etc/docker/daemon.json`.

   You should add the `insecure-registries` section to it. For example:

   ```
   {
     "debug": true,
     "experimental": false,
     "insecure-registries": [
       "REGISTRY_IP:5000"          <<< Replace REGISTRY_IP with actual IP from step 1
     ]
   }
   ```

   On MacOS if you're using Docker Desktop you can find this under 'Preferences' / 'Docker Engine'

   After changing this configuration you have to restart the Docker daemon.

> 4. Make sure you can push locally to Docker Registry

Reviewer 2: ```suggestion
4. Make sure you can push locally to your Docker Registry
```

   Execute the following, to make sure your Docker daemon can work with your Docker Registry

   ```
   docker pull gcr.io/google-samples/hello-app:1.0
   docker tag gcr.io/google-samples/hello-app:1.0 $REGISTRY_IP:$REGISTRY_PORT/hello-app:1.0
   docker push $REGISTRY_IP:$REGISTRY_PORT/hello-app:1.0
   ```

5. Start your Kind cluster to work with your Docker Registry

   Rather than simply starting your Kind cluster with `kind create cluster` use the following:

   ```
   export KIND_CLUSTER_NAME=kind

   cat << EOF | kind create cluster --name "${KIND_CLUSTER_NAME}" --config=-
   kind: Cluster
   apiVersion: kind.x-k8s.io/v1alpha4
   containerdConfigPatches: 
   - |-
     [plugins."io.containerd.grpc.v1.cri".registry.mirrors."$REGISTRY_IP:$REGISTRY_PORT"]
       endpoint = ["http://$REGISTRY_IP:$REGISTRY_PORT"]
   EOF
   ```

>    Note, how we use `http` in `endpoint` value, which makes Kind work with your 'insecure' registry.

Reviewer 2: ```suggestion
   Note, how we use `http` in the `endpoint` value, which makes Kind work with your 'insecure' registry.
```

   That can be done using the following command:

   ```
   sed -Ei -e "s#(image|value): strimzi/([a-z0-9-]+):latest#\1: ${DOCKER_REG}/strimzi/\2:latest#" \
       -e "s#([0-9.]+)=strimzi/([a-zA-Z0-9-]+:[a-zA-Z0-9.-]+-kafka-[0-9.]+)#\1=${DOCKER_REG}/strimzi/\2#" \
       install/cluster-operator/050-Deployment-strimzi-cluster-operator.yaml
   ```

   This will update `050-Deployment-strimzi-cluster-operator.yaml` replacing all the image references (in `image` and `value` properties) with ones with the same name from `$REGISTRY_IP:$REGISTRY_PORT/strimzi`.

9. Change the `RoleBindings` and `ClusterRoleBindinga` namespace to `default`

        sed -Ei -e 's/namespace: .*/namespace: default/' install/cluster-operator/*.yaml

   As an alternative you may want to create your own namespace, and use that.

   For example:
   
        kubectl create ns kafka
        sed -Ei -e 's/namespace: .*/namespace: kafka/' install/cluster-operator/*.yaml

   But, you then have to use `-n kafka` when issuing `kubectl` commands.
   
10. Then you can deploy the Cluster Operator by running:

        kubectl create -f install/cluster-operator

>     You can follow it's status by executing:

Reviewer 2: ```suggestion
    You can follow its status by executing:
```
